servlet-feature-pack/common/src/main/resources/modules/system/layers/base/org/jboss/as/naming/main/module.xml

```diff
@@ -51,5 +51,14 @@
         <!-- It is used only if org.wildfly.undertow.http-invoker capability is available -->
         <module name="io.undertow.core" optional="true"/>
         <module name="sun.jdk" />
+        <!--
+        The Angus Mail services are required to be able to load the Jakarta Mail Session from this Module context classloader
+        when listing the Mail context stored in the JNDI tree.
```

Comment:
Is this issue primarily about listing, or is the listing aspect just a detail of a test? IOW is the core problem here that using the provider doesn't work? Or is it just that the list of Provider objects (which afaict are just metadata holders) from Session.getProviders() is inaccurate?

Reply:
Hi @bstansberry , maybe showing the error trace I can make it more clear what the issue is. We get this error trace when listing the JNDI view using the naming subsystem:

`[standalone@localhost:9990 /] /subsystem=naming:jndi-view()`

The error is:

```
16:30:28,591 ERROR [org.jboss.as.controller.management-operation] (management-handler-thread - 1) WFLYCTL0013: Operation ("jndi-view") failed - address: ([("subsystem" => "naming")]): java.lang.IllegalStateException: Not provider of jakarta.mail.util.StreamProvider was found
    at jakarta.mail.api@2.1.1.jbossorg-1//jakarta.mail.util.FactoryFinder.find(FactoryFinder.java:64)
    at jakarta.mail.api@2.1.1.jbossorg-1//jakarta.mail.util.StreamProvider.provider(StreamProvider.java:186)
    at jakarta.mail.api@2.1.1.jbossorg-1//jakarta.mail.Session.<init>(Session.java:254)
    at jakarta.mail.api@2.1.1.jbossorg-1//jakarta.mail.Session.getInstance(Session.java:308)
    at org.jboss.as.mail@28.0.0.Beta1-SNAPSHOT//org.jboss.as.mail.extension.SessionProviderFactory$ManagedSession.getSession(SessionProviderFactory.java:171)
    at org.jboss.as.mail@28.0.0.Beta1-SNAPSHOT//org.jboss.as.mail.extension.MailSessionService.getValue(MailSessionService.java:73)
    at org.jboss.as.mail@28.0.0.Beta1-SNAPSHOT//org.jboss.as.mail.extension.MailSessionManagedReferenceFactory.getReference(MailSessionManagedReferenceFactory.java:56)
    at org.jboss.as.mail@28.0.0.Beta1-SNAPSHOT//org.jboss.as.mail.extension.MailSessionManagedReferenceFactory.getInstanceClassName(MailSessionManagedReferenceFactory.java:50)
    at org.jboss.as.naming@28.0.0.Beta1-SNAPSHOT//org.jboss.as.naming.ServiceBasedNamingStore.list(ServiceBasedNamingStore.java:189)
    at org.jboss.as.naming@28.0.0.Beta1-SNAPSHOT//org.jboss.as.naming.NamingContext.list(NamingContext.java:354)
    at org.jboss.as.naming@28.0.0.Beta1-SNAPSHOT//org.jboss.as.naming.NamingContext.list(NamingContext.java:370)
    at org.jboss.as.naming@28.0.0.Beta1-SNAPSHOT//org.jboss.as.naming.management.JndiViewOperation.addEntries(JndiViewOperation.java:122)
    at org.jboss.as.naming@28.0.0.Beta1-SNAPSHOT//org.jboss.as.naming.management.JndiViewOperation.addEntries(JndiViewOperation.java:136)
    at org.jboss.as.naming@28.0.0.Beta1-SNAPSHOT//org.jboss.as.naming.management.JndiViewOperation$1.execute(JndiViewOperation.java:72)
    at org.jboss.as.controller@20.0.0.Beta3//org.jboss.as.controller.AbstractOperationContext.executeStep(AbstractOperationContext.java:1045)
    at org.jboss.as.controller@20.0.0.Beta3//org.jboss.as.controller.AbstractOperationContext.processStages(AbstractOperationContext.java:777)
    at org.jboss.as.controller@20.0.0.Beta3//org.jboss.as.controller.AbstractOperationContext.executeOperation(AbstractOperationContext.java:466)
    at org.jboss.as.controller@20.0.0.Beta3//org.jboss.as.controller.OperationContextImpl.executeOperation(OperationContextImpl.java:1433)
    at org.jboss.as.controller@20.0.0.Beta3//org.jboss.as.controller.ModelControllerImpl.internalExecute(ModelControllerImpl.java:449)
    at org.jboss.as.controller@20.0.0.Beta3//org.jboss.as.controller.ModelControllerImpl.lambda$executeForResponse$0(ModelControllerImpl.java:260)
    at org.wildfly.security.elytron-base@2.0.0.Final//org.wildfly.security.auth.server.SecurityIdentity.runAs(SecurityIdentity.java:304)
    at org.wildfly.security.elytron-base@2.0.0.Final//org.wildfly.security.auth.server.SecurityIdentity.runAs(SecurityIdentity.java:270)
    at org.jboss.as.controller@20.0.0.Beta3//org.jboss.as.controller.ModelControllerImpl.executeForResponse(ModelControllerImpl.java:260)
    at org.jboss.as.controller@20.0.0.Beta3//org.jboss.as.controller.ModelControllerImpl.executeOperation(ModelControllerImpl.java:254)
    at org.jboss.as.controller@20.0.0.Beta3//org.jboss.as.controller.ModelControllerImpl.execute(ModelControllerImpl.java:237)
    at org.jboss.as.controller@20.0.0.Beta3//org.jboss.as.controller.remote.ModelControllerClientOperationHandler$ExecuteRequestHandler.doExecute(ModelControllerClientOperationHandler.java:241)
    at org.jboss.as.controller@20.0.0.Beta3//org.jboss.as.controller.remote.ModelControllerClientOperationHandler$ExecuteRequestHandler$1$1.run(ModelControllerClientOperationHandler.java:163)
    at org.jboss.as.controller@20.0.0.Beta3//org.jboss.as.controller.remote.ModelControllerClientOperationHandler$ExecuteRequestHandler$1$1.run(ModelControllerClientOperationHandler.java:159)
    at org.wildfly.security.elytron-base@2.0.0.Final//org.wildfly.security.auth.server.SecurityIdentity.runAs(SecurityIdentity.java:328)
    at org.wildfly.security.elytron-base@2.0.0.Final//org.wildfly.security.auth.server.SecurityIdentity.runAs(SecurityIdentity.java:285)
    at org.jboss.as.controller@20.0.0.Beta3//org.jboss.as.controller.AccessAuditContext.doAs(AccessAuditContext.java:254)
    at org.jboss.as.controller@20.0.0.Beta3//org.jboss.as.controller.AccessAuditContext.doAs(AccessAuditContext.java:225)
    at org.jboss.as.controller@20.0.0.Beta3//org.jboss.as.controller.remote.ModelControllerClientOperationHandler$ExecuteRequestHandler$1.execute(ModelControllerClientOperationHandler.java:159)
    at org.jboss.as.protocol@20.0.0.Beta3//org.jboss.as.protocol.mgmt.ManagementRequestContextImpl$1.doExecute(ManagementRequestContextImpl.java:70)
    at org.jboss.as.protocol@20.0.0.Beta3//org.jboss.as.protocol.mgmt.ManagementRequestContextImpl$AsyncTaskRunner.run(ManagementRequestContextImpl.java:160)
    at org.jboss.threads@2.4.0.Final//org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
    at org.jboss.threads@2.4.0.Final//org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1990)
    at org.jboss.threads@2.4.0.Final//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
    at org.jboss.threads@2.4.0.Final//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377)
    at java.base/java.lang.Thread.run(Thread.java:829)
    at org.jboss.threads@2.4.0.Final//org.jboss.threads.JBossThread.run(JBossThread.java:513)
```

The main problem is any caller module that needs to load a Mail Session Object now needs to see the Angus Mail services because in order to get a Session instance, we need to have an implementation of `StreamProvider`, which is provided by Angus Mail.

Firstly, we fixed it by explicitly using the "jakarta.mail.api" module as the TCCL, but that solution left out the other classloaders where the user can add other mail providers.

Now, since WildFly needs to load the session object when the naming subsystem lists the JNDI thread, we have added the Angus Mail services to the `org.jboss.as.naming` module.

Another third solution, which could be more reliable, could be to use a combined class loader when we are getting the session. For example a combination between the caller classloader and the "jakarta.mail.api" classloader.

Reply:
Thanks. I understood the reported problem; what I wasn't clear about is whether this was a symptom of a much bigger problem. We want to fix the reported problem, of course, and I'm happy to merge this to fix that problem. But jndi-view is a fairly minor feature so I want to understand if there is a bigger issue that affect application behavior, not just the management layer.

Reply:
In my opinion, there should not be an issue with the application behavior, since both, the Jakarta Mail API and the Angus Mail implementation, are dependencies visible to the application classpath.